### ISSUE TITLE
Feature Request - Incomplete restore

### ISSUE THREAD

**Issue by Komport** [ShahriyarR/MySQL-AutoXtraBackup#176]
Hi Muallim

Is there any way to restore my backups to the some point. As I came from Oracle and SQL Server I will try to explain with their feature. 
Say in Oracle we have Point in Time Recovery. We can say restore my database to this point. Or in SQL Server we are taking Transnational Backups, and we can restore all of them or few of them. 
Can we implement same in autoxtrabackup? Say if I took Full + 5 INC backups and I want to restore FULL + 3 INC.  Even better if we can choose TAG to which we want to restore.

**Comment by ShahriyarR** [comment]
Hi,
Even you can do the incomplete restore we can not  name it as PITR(point-in-time-recovery).
This is because:
* You took 1 full + 3 incrementals at 14:00.
* The crash happened at 15:00.
* You have restored from backup.
* Now you have to apply binary log entries to go to crash point(for this 1 hour interval).
XtraBackup itself do not copy binary logs to backup folder, because it is on user to keep binary logs or not. XB's duty is to backup tables. The reason I don't want to copy binary logs, is that the logs can be in huge amount of size and it will affect the tool's overall performance. So to be clear, applying binary logs should be done by user.

But I can implement those things from here:
* To prepare/recovery up to given number of increments.
* To prepare/recovery up to given TAG.

Are you okay with this?

**Comment by Komport** [comment]
Yes Muallim. I have mentioned PITR to better explain my point. This all about what you have pointed to. Restoring by TAG will be awesome feature.

**Comment by ShahriyarR** [comment]
It is really cool feature, thanks for this request.
I will implement this for 1.5.0 version of autoxtrabackup.

**Comment by ShahriyarR** [comment]
The first tests:

```
$ sudo autoxtrabackup --show_tags  --defaults_file=/home/shahriyar.rzaev/XB_TEST/server_dir/xb_2_3_ps_5_6.conf
Backup                  Type    Status  TAG
-------------------------------------------
2017-11-09_19-37-16     Full    OK      'This the full backup'
2017-11-09_19-38-03     Inc     OK      'This 1 INC backup'
2017-11-09_19-39-00     Inc     OK      'This 2 INC backup'
2017-11-09_19-40-08     Inc     OK      'This 3 INC backup'
```

* Passing Full backup tag name:

```
$ autoxtrabackup --tag='This the full backup' --defaults_file=/home/shahriyar.rzaev/XB_TEST/server_dir/xb_2_3_ps_5_6.conf -lf /home/shahriyar.rzaev/autoxtrabackup2.log --prepare -v -l DEBUG
2017-11-09 21:43:10 DEBUG    <pid.PidFile object at 0x7fdb9556c728> entering setup
2017-11-09 21:43:10 DEBUG    <pid.PidFile object at 0x7fdb9556c728> create pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-09 21:43:10 DEBUG    <pid.PidFile object at 0x7fdb9556c728> check pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Preparing full/inc backups!
What do you want to do?
1. Prepare Backups and keep for future usage. NOTE('Once Prepared Backups Can not be prepared Again')
2. Prepare Backups and restore/recover/copy-back immediately
3. Just copy-back previously prepared backups
Please Choose one of options and type 1 or 2 or 3: 1

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
2017-11-09 21:43:14 DEBUG    - - - - Preparing Full Backup - - - -
2017-11-09 21:43:14 DEBUG    Using xbstream to extract and decrypt from full_backup.stream!
2017-11-09 21:43:14 DEBUG    The following xbstream command will be executed /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xbstream -x --parallel=100 --decrypt=AES256 --encrypt-key=VVTBwgM4UhwkTTV98fhuj+D1zyWoA89K --encrypt-threads=4 < /home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_19-37-16/full_backup.stream -C /home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_19-37-16
2017-11-09 21:43:15 DEBUG    OK: XBSTREAM command succeeded.
2017-11-09 21:43:15 DEBUG    Trying to decrypt backup
2017-11-09 21:43:15 DEBUG    Running decrypt command -> /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup --decrypt=AES256 --encrypt-key=VVTBwgM4UhwkTTV98fhuj+D1zyWoA89K --target-dir=/home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_19-37-16 --remove-original
2017-11-09 21:43:15 DEBUG    1109 21:43:15 completed OK!
2017-11-09 21:43:15 DEBUG    OK: Decrypted!
2017-11-09 21:43:15 DEBUG    Trying to decompress backup
2017-11-09 21:43:15 DEBUG    Running decompress command -> /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup --decompress=TRUE --target-dir=/home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_19-37-16 --remove-original
2017-11-09 21:43:16 DEBUG    1109 21:43:16 completed OK!
2017-11-09 21:43:16 DEBUG    OK: Decompressed
2017-11-09 21:43:16 DEBUG    Running prepare command -> /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup --prepare --target-dir=/home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_19-37-16 --no-version-check --core-file
2017-11-09 21:43:47 DEBUG    /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup version 2.3.8 based on MySQL server 5.6.24 Linux (x86_64) (revision id: )
xtrabackup: cd to /home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_19-37-16/
xtrabackup: This target seems to be not prepared yet.
xtrabackup: xtrabackup_logfile detected: size=2097152, start_lsn=(58447684)
InnoDB: The universal page size of the database is set to 4096.
xtrabackup: using the following InnoDB configuration for recovery:
xtrabackup:   innodb_data_home_dir = ./
xtrabackup:   innodb_data_file_path = ibdata1:12M:autoextend
xtrabackup:   innodb_log_group_home_dir = ./
xtrabackup:   innodb_log_files_in_group = 1
xtrabackup:   innodb_log_file_size = 2097152
InnoDB: The universal page size of the database is set to 4096.
xtrabackup: using the following InnoDB configuration for recovery:
xtrabackup:   innodb_data_home_dir = ./
xtrabackup:   innodb_data_file_path = ibdata1:12M:autoextend
xtrabackup:   innodb_log_group_home_dir = ./
xtrabackup:   innodb_log_files_in_group = 1
xtrabackup:   innodb_log_file_size = 2097152
xtrabackup: Starting InnoDB instance for recovery.
xtrabackup: Using 104857600 bytes for buffer pool (set by --use-memory parameter)
InnoDB: Using atomics to ref count buffer pool pages
2017-11-09 21:43:16 7f8d41617840 InnoDB: !!!!!!!! UNIV_DEBUG switched on !!!!!!!!!
InnoDB: The InnoDB memory heap is disabled
InnoDB: Mutexes and rw_locks use GCC atomic builtins
InnoDB: Memory barrier is not used
InnoDB: Compressed tables use zlib 1.2.7
InnoDB: Using CPU crc32 instructions
InnoDB: Initializing buffer pool, size = 100.0M
InnoDB: Completed initialization of buffer pool
InnoDB: Highest supported file format is Barracuda.
InnoDB: Log scan progressed past the checkpoint lsn 58447684
InnoDB: Database was not shutdown normally!
InnoDB: Starting crash recovery.
InnoDB: Reading tablespace information from the .ibd files...
InnoDB: Restoring possible half-written data pages 
InnoDB: from the doublewrite buffer...
InnoDB: Doing recovery: scanned up to log sequence number 58447716 (0%)
InnoDB: 128 rollback segment(s) are active.
InnoDB: Waiting for purge to start
InnoDB: 5.6.24 started; log sequence number 58447716
xtrabackup: Last MySQL binlog file position 191, file name mysql-bin.000007

xtrabackup: starting shutdown with innodb_fast_shutdown = 1
InnoDB: FTS optimize thread exiting.
InnoDB: Starting shutdown...
InnoDB: Shutdown completed; log sequence number 58450404
InnoDB: The universal page size of the database is set to 4096.
xtrabackup: using the following InnoDB configuration for recovery:
xtrabackup:   innodb_data_home_dir = ./
xtrabackup:   innodb_data_file_path = ibdata1:12M:autoextend
xtrabackup:   innodb_log_group_home_dir = ./
xtrabackup:   innodb_log_files_in_group = 2
xtrabackup:   innodb_log_file_size = 1073741824
InnoDB: Using atomics to ref count buffer pool pages
2017-11-09 21:43:18 7f8d41617840 InnoDB: !!!!!!!! UNIV_DEBUG switched on !!!!!!!!!
InnoDB: The InnoDB memory heap is disabled
InnoDB: Mutexes and rw_locks use GCC atomic builtins
InnoDB: Memory barrier is not used
InnoDB: Compressed tables use zlib 1.2.7
InnoDB: Using CPU crc32 instructions
InnoDB: Initializing buffer pool, size = 100.0M
InnoDB: Completed initialization of buffer pool
InnoDB: Setting log file ./ib_logfile101 size to 1024 MB
InnoDB: Progress in MB: 100 200 300 400 500 600 700 800 900 1000
InnoDB: Setting log file ./ib_logfile1 size to 1024 MB
InnoDB: Progress in MB: 100 200 300 400 500 600 700 800 900 1000
InnoDB: Renaming log file ./ib_logfile101 to ./ib_logfile0
InnoDB: New log files created, LSN=58450404
InnoDB: Highest supported file format is Barracuda.
InnoDB: 128 rollback segment(s) are active.
InnoDB: Waiting for purge to start
InnoDB: 5.6.24 started; log sequence number 58450444
xtrabackup: starting shutdown with innodb_fast_shutdown = 1
InnoDB: FTS optimize thread exiting.
InnoDB: Starting shutdown...
InnoDB: Shutdown completed; log sequence number 58450454
171109 21:43:47 completed OK!
2017-11-09 21:43:47 DEBUG    - - - - The end of the Prepare Stage. - - - -
2017-11-09 21:43:47 DEBUG    <pid.PidFile object at 0x7fdb9556c728> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
2017-11-09 21:43:47 DEBUG    <pid.PidFile object at 0x7fdb9556c728> closing pidfile: /tmp/MySQL-AutoXtraBackup/autoxtrabackup.pid
```

**Comment by ShahriyarR** [comment]
Backup commands:

```
autoxtrabackup -v -lf /home/shahriyar.rzaev/autoxtrabackup2.log -l DEBUG --defaults_file=/home/shahriyar.rzaev/XB_TEST/server_dir/xb_2_3_ps_5_6.conf --backup --tag="This the full backup"

autoxtrabackup -v -lf /home/shahriyar.rzaev/autoxtrabackup2.log -l DEBUG --defaults_file=/home/shahriyar.rzaev/XB_TEST/server_dir/xb_2_3_ps_5_6.conf --backup --tag="This 1 INC backup"

autoxtrabackup -v -lf /home/shahriyar.rzaev/autoxtrabackup2.log -l DEBUG --defaults_file=/home/shahriyar.rzaev/XB_TEST/server_dir/xb_2_3_ps_5_6.conf --backup --tag="This 2 INC backup"

autoxtrabackup -v -lf /home/shahriyar.rzaev/autoxtrabackup2.log -l DEBUG --defaults_file=/home/shahriyar.rzaev/XB_TEST/server_dir/xb_2_3_ps_5_6.conf --backup --tag="This 3 INC backup"
```

**Comment by ShahriyarR** [comment]
* Passing 1st Incremental tag.
Result: FAILED

```
$ autoxtrabackup --tag='This 1 INC backup' --defaults_file=/home/shahriyar.rzaev/XB_TEST/server_dir/xb_2_3_ps_5_6.conf -lf /home/shahriyar.rzaev/autoxtrabackup2.log --prepare -v -l DEBUG

2017-11-09 22:25:25 ERROR    FAILED: Incremental BACKUP prepare
2017-11-09 22:25:25 ERROR    /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup version 2.3.8 based on MySQL server 5.6.24 Linux (x86_64) (revision id: )
incremental backup from 58447776 is enabled.
xtrabackup: cd to /home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_22-11-03/
xtrabackup: This target seems to be already prepared.
```
But to be honest could not find the reason:
Full backup prepared with:

```
2017-11-09 22:24:26 DEBUG    Running prepare command -> /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup --prepare --apply-log-only --target-dir=/home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_22-11-03 --no-version-check --core-file
```

Incremental prepared with:
```
2017-11-09 22:24:41 DEBUG    Running prepare command -> /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup --prepare --target-dir=/home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_22-11-03 --incremental-dir=/home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/inc/2017-11-09_22-11-37 --no-version-check --core-file
2017-11-09 22:25:23 DEBUG    /home/shahriyar.rzaev/XB_TEST/server_dir/target/percona-xtrabackup-2.3.x-debug/bin/xtrabackup version 2.3.8 based on MySQL server 5.6.24 Linux (x86_64) (revision id: )
incremental backup from 58447769 is enabled.
xtrabackup: cd to /home/shahriyar.rzaev/XB_TEST/backup_dir/ps_5_6_x_2_3/full/2017-11-09_22-11-03/
xtrabackup: This target seems to be already prepared with --apply-log-only.
xtrabackup: xtrabackup_logfile detected: size=2097152, start_lsn=(58447769)
InnoDB: The universal page size of the database is set to 4096.
```

So they are seems to be valid.

**Comment by ShahriyarR** [comment]
Passing 2nd Incremental tag and and 3rd as well succeeded.

**Comment by ShahriyarR** [comment]
So related to this feature I have implemented to prepare until given tag name.
The --tag option is universal here in term of:
* If it is passed with --backup option it will create given backup tag
* If it is passed with --prepare option will search given tag name from backup tags and prepare until this tag name, included.

This is wonderful thing actually.
But I have discussed with other world community members and decided not to implement:
* To prepare/recovery up to given number of increments.

Because could not find practical usage if there are no tags. In other words, getting backup with tags and preparing with tags is sufficient at this moment.
